### ISSUE TITLE
MFC libs missing in Visual Studio 2022

### ISSUE THREAD

**Issue by adlud** [actions/virtual-environments#5125]
### Description

With the move to Windows Server 2022 and Visual Studio 2022, my build job fails with this error:
2>C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Microsoft\VC\v160\Microsoft.CppBuild.targets(479,5): error MSB8041: MFC libraries are required for this project. Install them from the Visual Studio installer (Individual Components tab) for any toolsets and architectures being used.

This build job ran successfully with Visual Studio 2019. My native build environment is Visual Studio 2019, so maybe I need to move to VS 2022 natively, but the error message above makes me think that the MFC libs for VS 2022 need to be installed in the GitHub Actions environment.

### Virtual environments affected

- [ ] Ubuntu 18.04
- [ ] Ubuntu 20.04
- [ ] macOS 10.15
- [ ] macOS 11
- [ ] Windows Server 2016
- [ ] Windows Server 2019
- [X] Windows Server 2022

### Image version and build link

20220220.1

https://github.com/averydennison/pcm_FileRetriever/actions/runs/1887452708

Don't know if you can open that link, so let me know if you need more info.

### Is it regression?

20220131.1  https://github.com/averydennison/pcm_FileRetriever/actions/runs/1861017906

### Expected behavior

Build should succeed.

### Actual behavior

Build fails with error message given above.

### Repro steps

Run a job on Windows Server 2022 and Visual Studio 2022 that includes building a Visual C++ project that uses MFC.

**Comment by nikolai-frolov** [comment]
Hello @adlud @mavriq-dev we can add required VS components to Windows 2022 images but we need to know what exactly components are required for your project. You can compare [Windows 2019 VS components](https://github.com/actions/virtual-environments/blob/main/images/win/Windows2019-Readme.md#workloads-components-and-extensions) and [Windows 2022 VS components](https://github.com/actions/virtual-environments/blob/main/images/win/Windows2022-Readme.md#workloads-components-and-extensions) to find difference.

This snippet should install the component in runtime:
```
Set-Location "C:\Program Files (x86)\Microsoft Visual Studio\Installer\"
$InstallPath = "C:\Program Files\Microsoft Visual Studio\2022\Enterprise"
$componentsToAdd = @(
  "Microsoft.VisualStudio.Component.VC.v141.MFC.ARM.Spectre"
)
[string]$workloadArgs = $componentsToAdd | ForEach-Object {" --add " +  $_}
$Arguments = ('/c', "vs_installer.exe", 'modify', '--installPath', "`"$InstallPath`"",$workloadArgs, '--quiet', '--norestart', '--nocache')
$process = Start-Process -FilePath cmd.exe -ArgumentList $Arguments -Wait -PassThru -WindowStyle Hidden
if ($process.ExitCode -eq 0)
{
    Write-Host "components have been successfully added"
}
else
{
    Write-Host "components were not installed"
    exit 1
}
```

**Comment by adlud** [comment]
Hi Nikolay,
As best I can tell, these are the components that I need added:

Microsoft.VisualStudio.Component.VC.ATL | 17.1.32112.364
Microsoft.VisualStudio.Component.VC.ATLMFC | 17.1.32112.364
Microsoft.VisualStudio.Component.VC.v141.ATL | 17.1.32112.364
Microsoft.VisualStudio.Component.VC.v141.MFC | 17.1.32112.364

Thanks,
Dave L.

**Comment by miketimofeev** [comment]
@adlud I'm afraid we don't have access to your repo. Could you share some public code example that requires the libraries? Another point — it would be great if you can search through the available components and find the right ones (there are few of them that contains MFC in their name) https://docs.microsoft.com/en-us/visualstudio/install/workload-component-id-vs-enterprise?view=vs-2022

**Comment by KrisWragg** [comment]
The 2022 toolset should be covered by the package `Microsoft.VisualStudio.Component.VC.ATLMFC` that is already installed.

Currently we are transitioning from 2019 to 2022 toolset so I have been manually modifying 'toolset-2022.json' to add additional packages. I think you need to do the same @adlud ?

After "wasm.tools" I've got the following, you likely need to do the same, although your example is v141 so I assume you've still got projects targeting VS 2017 toolset?

```
"Microsoft.VisualStudio.ComponentGroup.VC.Tools.142.x86.x64",
"Microsoft.VisualStudio.Component.VC.14.29.16.11.ATL",
"Microsoft.VisualStudio.Component.VC.14.29.16.11.MFC",
"Microsoft.VisualStudio.Component.VC.14.29.16.11.CLI.Support",	
```

**Comment by mavriq-dev** [comment]
I found my issue. The platform toolset in my vsproj was set for 2019 so it couldn't find the correct version of MFC. My bad.

**Comment by adlud** [comment]
@miketimofeev

I will sanitize the repo I am using and copy it to a public one. Will get back to you with the name.

From the link you provided, I found these relevant components:
Microsoft.VisualStudio.Component.VC.ATLMFC
Microsoft.VisualStudio.Component.VC.14.29.16.11.MFC
Microsoft.VisualStudio.Component.VC.14.29.16.11.MFC.Spectre
Microsoft.VisualStudio.Component.VC.14.30.17.0.MFC
Microsoft.VisualStudio.Component.VC.14.30.17.0.MFC.Spectre
Microsoft.VisualStudio.Component.VC.ATLMFC.Spectre
Microsoft.VisualStudio.Component.VC.v141.MFC
Microsoft.VisualStudio.Component.VC.v141.MFC.Spectre
Microsoft.VisualStudio.Component.VC.ATL
Microsoft.VisualStudio.Component.VC.14.29.16.11.ATL
Microsoft.VisualStudio.Component.VC.14.29.16.11.ATL.Spectre
Microsoft.VisualStudio.Component.VC.14.30.17.0.ATL
Microsoft.VisualStudio.Component.VC.14.30.17.0.ATL.Spectre
Microsoft.VisualStudio.Component.VC.ATL.Spectre
Microsoft.VisualStudio.Component.VC.v141.ATL
Microsoft.VisualStudio.Component.VC.v141.ATL.Spectre

**Comment by adlud** [comment]
@KrisWragg 
As you say, Microsoft.VisualStudio.Component.VC.ATLMFC may cover some of the other components I listed. I don't think it will cover Microsoft.VisualStudio.Component.VC.ATL, Microsoft.VisualStudio.Component.VC.ATLMFC.Spectre,  or Microsoft.VisualStudio.Component.VC.ATL.Spectre. With all 4 of those installed, all the components may be covered, but I am not sure about that.

I am stuck on VS 2019 natively so will continue to target that in my VC++ project. The v141 in the component names came from the links others have posted.

**Comment by adlud** [comment]
@miketimofeev
I have created the public repo adlud/mfcBuildTest .
Based on comments from mavriq-dev, probably the first thing you should try is changing the VC++ project settings to target the VS 2022 toolset. If the project then builds successfully, the only thing more to do is me closing this issue.  Thanks.

**Comment by dsame** [comment]
@adlud Upgrading toolset from 142 to 143 solves the issue: 
https://github.com/akv-platform/mfcBuildTest/runs/5381397549?check_suite_focus=true

I believe this is the right to fix the problem and now frameworks of the older versions need to be installed on VS2022 image.

Do you think the issue can be closed

**Comment by adlud** [comment]
@dsame Yes, I will go ahead and close the issue. Thanks for confirming that the project builds successfully on the Windows Server 2022 image when using a matching toolset version in the VC++ project.

When you say that "frameworks of the older versions need to be installed on VS2022 image", will doing so allow the toolset in the VC++ project to be kept at 142 and be successfully built on the Windows Server 2022 image?
